### PR TITLE
RPC API miner_getstathr: Add "ispaused" information into response.

### PR DESCRIPTION
**Pull request by StefanOberhumer**
Adding possibility to check ispaused if hashrate on a card drops to 0.

### CODE REVIEW

File: libapicore/ApiServer.cpp [hunk 1]
```diff
@@ -423,6 +424,13 @@ Json::Value ApiConnection::getMinerStatHR()
 		gpuIndex++;
 	}
 
+	gpuIndex = 0;
```
Comment (AndreaLanfranchi) [inline libapicore/ApiServer.cpp]: Instead of having 3 different for loops (inlcuding this last yours) can we manage to pull info from all gpus using only 1 for loop ?

Comment (AndreaLanfranchi) [reply libapicore/ApiServer.cpp]: Also ... ispaused is a little too generic. 
Maybe we could add some info about the pausing reason ? (eg. temp threshold )

Comment (StefanOberhumer) [reply libapicore/ApiServer.cpp]: loops :+1: 

ispaused: I already thought about that and came due following reasons to the current solution:
* as I implemented -tstop/-tstart I already thought about a pause bit flag (TEMPERATURE, USER, ...)
but currently we have only pausing due temperature reason (no keypress and no api to pause mining)
* ispaused is a bool, if there will be more reasons I thought the API can respond a "pausereasons" or so. I've noticed content strings changes more often than bools ;-) which always hurts in interfaces.

So I have following questions before adapting PR:
* ispaused: bool or string
* If ispaused is a string: when not paused: null or "" or "no" or "0" ?
* If ispaused is a string and if multiple reasons for pausing pending - join reasons with "," or ";"
* Should I prepare pausing to be a unsigned int with bits and adapt is_mining_paused() ?

Comment (AndreaLanfranchi) [reply libapicore/ApiServer.cpp]: I'll try explain what I would like to achieve.
At present the result of getstathr is like this

```
{  
   "id":0,
   "jsonrpc":"2.0",
   "result":{  
      "ethhashrate":87388615,
      "ethhashrates":[  
         14506510,
         14593898,
         14593898,
         14593898,
         14506510,
         14593898
      ],
      "ethinvalid":0,
      "ethpoolsw":0,
      "ethrejected":0,
      "ethshares":136,
      "fanpercentages":[  
         90,
         90,
         90,
         90,
         90,
         90
      ],
      "pooladdrs":"10.0.0.113:5551",
      "powerusages":[  
         0.0,
         0.0,
         0.0,
         0.0,
         0.0,
         0.0
      ],
      "runtime":"112",
      "temperatures":[  
         54,
         51,
         57,
         59,
         62,
         60
      ],
      "version":"0.15.0.dev11-96+commit.6b02ef2f"
   }
}
```

I whish we could manage gpu's as an array of objects (instead of parsing a string which has to be split into an array of values)

```
{  
   "id":0,
   "jsonrpc":"2.0",
   "result":{  
      "hashrate":87388615,
      "shares":136,
      "invalid":0,
	  "rejected":0,
      "poolsw":0,
      "pooladdrs":"10.0.0.113:5551",
      "runtime":"112",
      "version":"0.15.0.dev11-96+commit.6b02ef2f.dirty",
	  "gpus" : [
			{ "index" : 0,
			  "hashrate : 14506510,
			  "power" : 0.0,
			  "fan" : 90,
			  "temp" : 54,
			  "temp-stop" : 0,
			  "temp-start" : 0,
			  "status" : "<operational|paused|error>",
			  "reason" : "<null|at-temp|manual|error>"
			},
			{ "index" : 1,
			  "hashrate : 14506510,
			  "power" : 0.0,
			  "fan" : 90,
			  "temp" : 54,
			  "temp-stop" : 0,
			  "temp-start" : 0,
			  "status" : "<operational|paused|error>",
			  "reason" : "<null|at-temp|manual|error>"
			},
			{ "index" : 2,
			  "hashrate : 14506510,
			  "power" : 0.0,
			  "fan" : 90,
			  "temp" : 54
			  "temp-stop" : 0,
			  "temp-start" : 0,
			  "status" : "<operational|paused|error>",
			  "reason" : "<null|at-temp|manual|error>"
			},
			{ "index" : 3,
			  "hashrate : 14506510,
			  "power" : 0.0,
			  "fan" : 90,
			  "temp" : 54
			  "temp-stop" : 0,
			  "temp-start" : 0,
			  "status" : "<operational|paused|error>",
			  "reason" : "<null|at-temp|manual|error>"
			},
			{ "index" : 4,
			  "hashrate : 14506510,
			  "power" : 0.0,
			  "fan" : 90,
			  "temp" : 54
			  "temp-stop" : 0,
			  "temp-start" : 0,
			  "status" : "<operational|paused|error>",
			  "reason" : "<null|at-temp|manual|error>"
			},
			{ "index" : 5,
			  "hashrate : 14506510,
			  "power" : 0.0,
			  "fan" : 90,
			  "temp" : 54
			  "temp-stop" : 0,
			  "temp-start" : 0,
			  "status" : "<operational|paused|error>",
			  "reason" : "<null|at-temp|manual|error>"
			}
	  ]
	  
   }
}
```

Would not set bit-shifting values for reasons of paused as this would imply consumers of api to have a legend of values and also bit shifting may be affected by endianess.

I know this scheme is a bit wordy but is more Json style than the actual.

Obviously this is my view, not necessarily the same of other devs. Thus some more feedback may be needed.

P.s. Anticipating your question about compatibility with ... erm ... other miners : getstathr is ethminer's only ... so we can change almost everything leaving gestat1 as is.

Comment (AndreaLanfranchi) [reply libapicore/ApiServer.cpp]: In this way we could also each gpu object any property we may need (ex. Vendor, Compute level and so on) changing only the structure of the GPU object while leaving the "result" structure intact.

Comment (StefanOberhumer) [reply libapicore/ApiServer.cpp]: > Would not set bit-shifting values for reasons of paused as this would imply consumers of api to have a legend of values and also bit shifting may be affected by endianess.

I just thought preparing internal a flag for pausing (replace 'm_wait_for_tstart_temp' with 'mining_paused_flag' and adding MINIG_PAUSED_FLAG_WAIT_TSTART_TEMP=0x01, MINIG_PAUSED_FLAG_MANUAL = 0x02,  MINIG_PAUSED_FLAG_ERROR = 0x04, ...)

> P.s. Anticipating your question about compatibility with ... erm ... other miners : getstathr is ethminer's only ... so we can change almost everything leaving gestat1 as is.

Yes I know - In miner_getstathr I feel free to change nearly everything ;-)

>Obviously this is my view, not necessarily the same of other devs. Thus some more feedback may be needed.

Yep - I'll wait for more reviews...

**Comment by jean-m-cyr** [comment]
Why aren't we moving towards a more RESTful approach. Modern choices seem to be between REST and SOAP these days.

http://miner1/status - would return the global properties of the miner. Things like miner name, software version, # of miners, etc... (using JSON encoding in a websocket wrapper).

http://miner1/gpu3/status - would return GPU specific attributes for GPU 3 (for example). ie. power, temp., fan speed, etc...

Action commands could be implemented as follows:

http://miner1/shutdown
http:://miner1/reboot

http://miner1/gpu2/disable
http://miner1/gpu3/enable

Anyway, you get the idea...

**Comment by AndreaLanfranchi** [comment]
While SOAP is a protocol and REST is an architectural style (thus not very comparable) the main issue I see using http transport is **it's staless**: every request have to pass through all authentication stack (if any) and in case of complex operations may involve the need to cache session state. It' a huge overhead not count the continuos opening and closing of sockets.
While this approach may help interactive users to have remote control over their instance of ethminer (sort of) it does not fit well if bundled with remote monitoring tools where, for example, the simple presence of an opened keepalive'd socket may indicate the liveness of the remote end.

**Comment by jean-m-cyr** [comment]
I guess we'll just have to agree to disagree. It's a documented standard, compared to the undocumented non standard mess we have now.

**Comment by AndreaLanfranchi** [comment]
I think the problem should be split in two.

Mining operations where monitoring is made inside the local network and **remote** access to single rig. In first case I do not know anyone (but my contacts may be limited) who would like to drop socket/json format (it's way easier to implement and way faster) while in the latter I may agree with you that for hobbyst/small operations (max 4 to 5 rigs) a higher level of security may be appreciated if you plan to expose ethminer on the web.

**Comment by AndreaLanfranchi** [comment]
Documentation is another matter. We could make a mess even using REST if we do not document well both the entry points and payloads (both incoming and outgoing).

**Comment by AndreaLanfranchi** [comment]
>I find the notion of authentication on a statefull unencrypted channel a just a little bizarre. There is no way I'd ever let a router/firewall allow an unencrypted miner connection to be initiated from outside my local network.

It depends on how sensible are the informations which could be pulled from that connection. 
Actually with ethminer's api you can do very little.

**Comment by jean-m-cyr** [comment]
@AndreaLanfranchi Are you intending to release this PR. It's not clear what specific changes you are requesting, and a complete rewrite or reformat really belong in a separate PR.

**Comment by AndreaLanfranchi** [comment]
> Is it the combination of loops you're waiting for?

Yes ... but if needed I can withdraw my comment immediately.

**Comment by StefanOberhumer** [comment]
* Joined the loops as suggested.
* For the detailed information api proposed by @AndreaLanfranchi I'd like to create an extra PR.

**Comment by AndreaLanfranchi** [comment]
Please also add new member in API_DOCUMENTATION
